File: app/src/main/java/com/privat/contacts/data/cloud/model/UserNet.java
package com.privat.contacts.data.cloud.model;

import com.google.gson.annotations.SerializedName;

import org.jetbrains.annotations.NotNull;

public class UserNet {
    @NotNull
    private final int id;
    @NotNull
    private final String uid;
    @NotNull
    private final String password;
    @NotNull
    @SerializedName("first_name")
    private final String firstName;
    @NotNull
    @SerializedName("last_name")
    private final String lastName;
    @NotNull
    private final String username;
    @NotNull
    private final String email;
    @NotNull
    private final String avatar;
    @NotNull
    private final String gender;
    @NotNull
    @SerializedName("phone_number")
    private final String phoneNumber;
    @NotNull
    @SerializedName("social_insurance_number")
    private final String socialInsuranceNumber;
    @NotNull
    @SerializedName("date_of_birth")
    private final String dateOfBirth;
    @NotNull
    private final EmploymentNet employment;
    @NotNull
    private final AddressNet address;
    @NotNull
    @SerializedName("credit_card")
    private final CreditCardNet creditCard;
    @NotNull
    private final SubscriptionNet subscription;
    public UserNet(int id,
                   String uid,
                   String password,
                   String firstName,
                   String lastName,
                   String username,
                   String email,
                   String avatar,
                   String gender,
                   String phoneNumber,
                   String socialInsuranceNumber,
                   String dateOfBirth,
                   EmploymentNet employment,
                   AddressNet address,
                   CreditCardNet creditCard,
                   SubscriptionNet subscription) {
        this.id = id;
        this.uid = uid;
        this.password = password;
        this.firstName = firstName;
        this.lastName = lastName;
        this.username = username;
        this.email = email;
        this.avatar = avatar;
        this.gender = gender;
        this.phoneNumber = phoneNumber;
        this.socialInsuranceNumber = socialInsuranceNumber;
        this.dateOfBirth = dateOfBirth;
        this.employment = employment;
        this.address = address;
        this.creditCard = creditCard;
        this.subscription = subscription;
    }
    public <T> T map(Mapper<T> mapper) {
        return mapper.map(id,
                uid,
                password,
                firstName,
                lastName,
                username,
                email,
                avatar,
                gender,
                phoneNumber,
                socialInsuranceNumber,
                dateOfBirth,
                employment,
                address,
                creditCard,
                subscription);
    }
    public interface Mapper<T> {
        T map(
                int id,
                String uid,
                String password,
                String firstName,
                String lastName,
                String username,
                String email,
                String avatar,
                String gender,
                String phoneNumber,
                String socialInsuranceNumber,
                String dateOfBirth,
                EmploymentNet employment,
                AddressNet address,
                CreditCardNet creditCard,
                SubscriptionNet subscription
        );
    }
}

File: app/src/main/java/com/privat/contacts/data/cloud/model/UserNet.kt
package com.privat.contacts.data.cloud.model

import com.google.gson.annotations.SerializedName

class UserNet(private val id: Int,
              private val uid: String,
              private val password: String,
              @field:SerializedName("first_name") private val firstName: String,
              @field:SerializedName("last_name") private val lastName: String,
              private val username: String,
              private val email: String,
              private val avatar: String,
              private val gender: String,
              @field:SerializedName("phone_number") private val phoneNumber: String,
              @field:SerializedName("social_insurance_number") private val socialInsuranceNumber: String,
              @field:SerializedName("date_of_birth") private val dateOfBirth: String,
              private val employment: EmploymentNet,
              private val address: AddressNet,
              @field:SerializedName("credit_card") private val creditCard: CreditCardNet,
              private val subscription: SubscriptionNet) {
    fun <T> map(mapper: Mapper<T>): T {
        return mapper.map(id,
                uid,
                password,
                firstName,
                lastName,
                username,
                email,
                avatar,
                gender,
                phoneNumber,
                socialInsuranceNumber,
                dateOfBirth,
                employment,
                address,
                creditCard,
                subscription)
    }

    interface Mapper<T> {
        fun map(
                id: Int,
                uid: String,
                password: String,
                firstName: String,
                lastName: String,
                username: String,
                email: String,
                avatar: String,
                gender: String,
                phoneNumber: String,
                socialInsuranceNumber: String,
                dateOfBirth: String,
                employment: EmploymentNet,
                address: AddressNet,
                creditCard: CreditCardNet,
                subscription: SubscriptionNet
        ): T
    }
}
